[tanqinyong] iP

tanqinyong opened the pull request:
## Felix

> "The result is irrelevant because the effort was there." - Felix Lengyel ([Source](https://youtu.be/dQw4w9WgXcQ?si=_zdoBkTwpO4RXmcL))

Felix gives you the opportunity to get focused and organize. It's:
- text-based
- intuitive
- easy to use

All you have to do is:
1. download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **_FREE!_**

Features:

- [ ]   Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]   Reminders (coming soon)

```java    
public TaskList(ArrayList<Task> taskList) {
        this.taskList = taskList;
        count = taskList.size();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tanqinyong committed:
[Level-0] Changed chatbot name and greetings.

tanqinyong committed:
[Level-1] Updated Felix(Duke) to echo commands entered by the user and exit when user types bye.

tanqinyong committed:
[Level-2] Add ability to store text and display them using an array.

tanqinyong committed:
[Level-3] Mark as done: Added the ability to mark tasks as done and undone.
Additionally, added the tasklist class to hold the list of all tasks.
Abstracted out the relevant functionality to the tasklist and task class accordingly.

tanqinyong committed:
[Level-4] Added ToDo, Deadlines and Events subclasses.
1. Updated Duke, Task and TaskList accordingly so that all functionalities work with new subclasses
2. Updated toString methods

tanqinyong committed:
[A-TextUiTesting] Performed some UI tests using I/O redirection
1. Tested default greeting by chatbot
2. Tested creation of todos and deadlines

tanqinyong committed:
[Level-5] Handled some relevant errors, null pointer, index out of bounds, number format and illegal argument.

tanqinyong committed:
[Level-6] Added a delete function.
- Changed the data structure used from a primitive array to arraylist.
- Modified the relevant functions in tasklist accordingly.
- Edited duke to accomodate for new function and added exception handling for it.

tanqinyong committed:
Added Javadocs to document all of the code.

tanqinyong committed:
Add save functionality to Duke

Duke does not save to hard disk and instead reboots with an empty task list every time it runs.

Add in new functionality that allows Duke to save to hard disk every time the task list changes.
Also loads in the task list every time it runs.

Abstract out storage class to store these operations.

tanqinyong committed:
Merge branch 'branch-Level-7'

tanqinyong committed:
Modify duke to accept and store proper dates

Duke accepts all kinds of input for date fields.

This can mean that users can enter invalid dates which do not make sense.

Change all task classes to accept and store LocalDate for all date fields.

This allows us to store, format and validify date inputs correctly, and print them out in a more readable format.

tanqinyong committed:
Restructure code for better OOP design

The main Duke class have some common behaviours and repeated code with similar functionalities.

To enhance readability and maintainability, it's important to refactor the code to a more organized and modular OOP design.

Let's,
* Add a new Ui class that manages user interactions.
* Add a new Storage class that handles task file loading and saving
* Add a new Parser class that interprets user commands

This approach makes it easier to understand and maintai neach component. The classes encapsulate related functionality, allowing for easier maintenance in the future.

tanqinyong committed:
Refactor code base into multiple packages

All classes are not in a package and are put in the java folder together. This makes it hard to distinguish related classes and makes the code base rather disorganized.

Let's refactor the code into three main packages:
* 'duke': Houses the main program and TaskList class.
* 'task': Manages the child tasks and the parent Task class.
* 'util': Handles user interactions.

This provides a clear separation of concerns and makes the code base organized and easy to navigate.

tanqinyong committed:
Merge branch 'add-gradle-support' of https://github.com/tanqinyong/ip

tanqinyong committed:
Configure the project with Gradle

Using Gradle as a buld automation tool allows for smooth project and dependency management.

Let's add Gradle to our project.

tanqinyong committed:
Add JUnit tests to verify functionality of two non-trivial methods

To ensure the reliability and functionality of the above two functions, we test them accordingly with JUnit.

Let's produce unit tests for:
* 'Deadline.toString()'
* 'Parser.parseDeadline()'
* 'Parser.parseDate()'

tanqinyong committed:
Package app as an executable JAR for easy distribution

Currently the app is not easy to distribute as it requires a lot of setup on the users' end.

Let's package the Duke project as a JAR so users can simply download and run the program.

Note that this commit does not have any significant change. (Removed extra line form TaskList)

tanqinyong committed:
Remove some JavaDocs comments for getters and setters

There are some redundant comments for getters and setters in the Task class.

Let's remove the comments.

tanqinyong committed:
Add new Find Tasks feature

Users may find it difficult to search for tasks given that they have a long list of tasks.

Add new 'findTask' method in the TaskList class. This method allows users to search for tasks by providing a substring.

The implementation is straightforward and easy to maintain, iterating through the task list and checking if each task's description matches the substring.

tanqinyong committed:
Merge branch 'branch-A-JavaDoc'

tanqinyong committed:
Merge branch 'branch-Level-9'

jinhanfromNUS commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,175 @@
+                    String byDate = splitCommand[1];
+                    Deadline dl = new Deadline(taskDescription, parser.parseDate(byDate));
+                    taskList.addTask(dl, false);
+                } catch (IndexOutOfBoundsException | IllegalArgumentException | DateTimeParseException e) {
```
I like how you handle the exceptions, a lot of people neglected it!

jinhanfromNUS commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,81 @@
+ *
+ *  @author Tan Qin Yong
+ */
+public class Task {
```
Your task class is well created and designed.

jinhanfromNUS commented on src/main/java/util/Ui.java:
```diff
@@ -0,0 +1,115 @@
+                + "|  ___|__| (_)_  __\n"
+                + "| |_ / _ \\ | \\ \\/ /\n"
+                + "|  _|  __/ | |>  < \n"
+                + "|_|  \\___|_|_/_/\\_\\\n";
```
This Felix is awesome!

jinhanfromNUS commented on src/main/java/util/Ui.java:
```diff
@@ -0,0 +1,115 @@
+     * Prints a message for marking tasks as done.
+     */
+    public void printMark() {
+        System.out.println("Great job agent 47. Marking this task as DONE: ");
```
Love this :)

jinhanfromNUS reviewed (commented):
Wow, your code is pretty complete and cool.

Emberlynn-Loo commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,175 @@
+                try {
+                    int taskNo = Integer.parseInt(commandArr[1]);
+                    taskList.deleteTask(taskNo);
+                } catch (NumberFormatException | NullPointerException | IndexOutOfBoundsException e) {
```
I like that you handled and covered pretty much all the potential exceptions

Emberlynn-Loo commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,175 @@
+            String command = commandArr[0].toLowerCase();
+
+            switch (command) {
+            case "bye": {
```
I like that you specified your cases instead of just putting it as case 1,2,3 etc as most people do. Makes it easier for readibility

Emberlynn-Loo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+        count++;
+
+        if (!loadTask) {
+            System.out.println("Cool! Adding new task: ");
+            System.out.println(newTask.toString());
+            System.out.println("Now you have " + this.count
+                    + " tasks in your list.");
```
Consider breaking the print statements into a separate method for better separation of concerns, as you did for your duke task with ui.printLine()

```suggestion
            UI.printLoadTask;
```

Emberlynn-Loo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+        int matchCount = 1;
+        for (int i = 0; i < count; i++) {
+            Task currentTask = taskList.get(i);
+            if (currentTask.getDescription().contains(keyWord)) {
```
Maybe consider handling the case if no matches are found?

Emberlynn-Loo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+     * @param taskNo The task number to retrieve the task.
+     */
+    public void markNotDoneAtInd(int taskNo) {
+        Task currentTask = this.getTask(taskNo);
```
Maybe consider handling the case if when the task number provided is invalid?

Emberlynn-Loo commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return The string "task.Task" representing the type of the task.
+     */
+    public String getType() {
+        return "Task";
+    }
```
Could try returning a type specific to each subclass (ie. Event, Todo, etc) by overriding?

Emberlynn-Loo commented on src/main/java/util/Parser.java:
```diff
@@ -0,0 +1,73 @@
+     * @throws DateTimeParseException If the date string cannot be parsed.
+     */
+    public LocalDate parseDate(String date) throws DateTimeParseException {
+        return LocalDate.parse(date, DateTimeFormatter.ofPattern("yyyy-MM-dd"));
```
Could add a 'DateTimeParseException if the date string cannot be parsed

Emberlynn-Loo reviewed (commented):
Overall, your good looks great! Just some minor nitpicking. Your UI class could handle some more of the UI logic to separate your classes further. Could also handle more exceptions. Otherwise, your code is well-organised! great work :))

tanqinyong committed:
Add GUI using JavaFX for Duke

Prior to this commit, Duke did not have GUI interface and only ran on console.

This is not very intuitive or user-friendly, as users have to manually enter console commands to run Duke.

Let's:
* Implement JavaFX in our codebase
* Add Launcher class as our new entrypoint
* Add Main class to handle the main GUI components
* Add GUI package with DialogBox and MainWindow controller classes
* Add two GUI views for DialogBox and MainWindow

tanqinyong committed:
Merge branch 'branch-Level-10'

tanqinyong committed:
Add Assertions to TaskList class

Previously, there was no use of assertions
to verify code behaviour.

Let's add some assertions
to important methods in the TaskList
class.

Using assertions will help
verify code behaviour and detect
for bugs, improving error handling.

tanqinyong committed:
Abstract out parsing of commands in main class

The main class consisted of one very long method,
which parsed the user input and perform the operations
accordingly.

But this is not very readable and not up to
coding standard guidelines.

Let's:
* Create methods in Parser for each command
* Abstract out from main to new methods
* Create constructor for Duke to initialise
   utilities

With this, our code is much more modular
and readable.

tanqinyong committed:
Abstract out helper function for Storage methods

Both methods in Storage were very long
and unreadable. This may cause issues
as it will be hard to identify bugs in the
future.

Additionally, add javadocs for methods in Parser
as well as for methods in Storage.

Let's:
* Abstract out helper functions for
each of the methods
* Add javadocs accordingly

tanqinyong committed:
Merge pull request #2 from tanqinyong/branch-A-Assertions

Add Assertions to TaskList class

tanqinyong committed:
Merge pull request #3 from tanqinyong/branch-A-CodeQuality

Abstract out parsing of commands in main class

tanqinyong committed:
Implement the ability to create notes

Previously, the only functionality for Felix
was to manage tasks. This was limited as
users could only store tasks.

Let's:
* Add a new Note class
* Add new functions in Parser
to accept note commands
* Edit Felix to handle note
commands

tanqinyong committed:
Change name of program from Duke to Felix

The program was still named Duke,
this may confuse users as it is now Felix.

Let's change all instances of Duke to Felix
and change the classes/package names.

tanqinyong committed:
Create User Guide for Felix

Previously, there was no user guide
for Felix. This is not user friendly.

Let's create a user guide in our
docs folder!

tanqinyong committed:
Perform minor changes to product image

Changed the name of the product image,
for script detection.

Let's change the name of the product image.
